### ISSUE TITLE
有些图片抠图时，会返回一张透明的，这种怎么处理

### ISSUE THREAD

**Issue by hjj-lmx** [ZhengPeng7/BiRefNet#131]
![QifWMKbZ9Sd](https://github.com/user-attachments/assets/5fa348fd-f02a-47df-9f5d-d10fe44318d5)

**Comment by ZhengPeng7** [comment]
<img width="1583" alt="截屏2024-11-26 18 31 01" src="https://github.com/user-attachments/assets/ddd96b27-fcf6-4a02-9481-98c805147cf3">

你是指这个全白的预测结果吗? 这个是表示没有检测出结果, 因为BiRefNet用的数据集主要是针对主体进行自动分割的, 而这张图中没有前景/显著/...等各种类别的 **主体**, 因此没有检测出按理来说其实还是正确的哈.

**Comment by hjj-lmx** [comment]
> 没有检测出按理说其实还是对的哈。

如果我想加判断，去处理这种，应该怎么做，如果没有前景就抛一个异常

**Comment by ZhengPeng7** [comment]
嗯, 当然也是合理的. 就只是不推荐写在`refine_foreground`里😂, 毕竟它不太属于这里.

**Comment by hjj-lmx** [comment]
> 嗯, 当然也是合理的. 就只是不推荐写在`refine_foreground`里😂, 毕竟它不太属于这里.

那应该写在什么地方，我如果返回没有前景的，业务会报错

**Comment by ZhengPeng7** [comment]
之前吧, 因为如果没有前景物体, 就不用经过refine啦, 逻辑上是这样的吧.

**Comment by hjj-lmx** [comment]
def has_foreground(pred_pil, threshold=0.5, min_foreground_ratio=0.01):
    """
    判断分割图是否有前景
    :param pred_pil: PIL.Image 对象 (二值分割图，像素范围 [0, 1])
    :param threshold: 判定前景的像素值阈值 (默认 0.5)
    :param min_foreground_ratio: 前景像素比例阈值 (默认 1%)
    :return: 布尔值，True 表示有前景，False 表示无前景
    """
    # 转为 NumPy 数组
    pred_array = np.array(pred_pil)

    # 如果范围是 [0, 255]，归一化到 [0, 1]
    if pred_array.max() > 1:
        pred_array = pred_array / 255.0

    # 计算前景像素比例
    foreground_ratio = (pred_array > threshold).mean()

    # 判断是否有前景
    return foreground_ratio > min_foreground_ratio
    这么设置的比例可以吗？我是在这里调用的
        # Prediction
        with torch.no_grad():
            preds = self.birefnet(input_images)[-1].sigmoid().cpu()
        pred = preds[0].squeeze()
        pred_pil = transforms.ToPILImage()(pred)
        if not has_foreground(pred_pil):
            raise ValueError("未识别到可抠图的前景图像")
        # 取前景
        image_masked = refine_foreground(image, pred_pil)
        不知道设置的0.5和0.01是否合理

**Comment by ZhengPeng7** [comment]
大致是可以的. 当然`0.01`具体得看你的情况, 如果需要调整, 也可以设置像素数量绝对值作为阈值.
